### ISSUE TITLE
Suggestion: polling loop instead of delay for boot / ssh keys

### ISSUE THREAD

**Issue by mattdm** [kushaldas/tunir#51]
In `tunirmultihost.py` when setting up a system, it currently does 

                print("We will wait for 45 seconds for the image to boot up.")
                time.sleep(45)

It's already a fundamental premise that we'll ssh in to run tests, right? What about rather than waiting for 45 seconds, doing a loop to test for port 22 being open? (Or maybe that port responding with "`SSH-.*`", or maybe using paramiko to _try_ to connect but expecting to get `BadHostKeyException` since that's not set up yet...) This loop would have only a short wait between tries (1, 2, 5 seconds?) and could have a higher timeout.

This would result in less delay in the optimal case, and at the same time, allow more time for the host to become ready if need be.

**Comment by kushaldas** [comment]
Correct, I have a ticket opened for this https://github.com/kushaldas/tunir/issues/36

You can see the same in use in the new gotun tool, for example:
http://status.kushaldas.in/job/gotun/6/console has the example of polling.

**Comment by kushaldas** [comment]
It seems I will still have to keep the initial 45 seconds delay. We get the IP of the system only after it boots up from arp table.

**Comment by mattdm** [comment]
> It seems I will still have to keep the initial 45 seconds delay. We get the IP of the system only after it boots up from arp table.

Hmmm. Doesn't that just mean doing _two_ loop tests — one for the arp table, followed by one for ssh?

**Comment by kushaldas** [comment]
That can be done, but it should be a special case, as many times the image does not boot at all, and we will be stuck at the finding IP step.

**Comment by mattdm** [comment]
> That can be done, but it should be a special case, as many times the image does not boot at all, and we will be stuck at the finding IP step.

Sorry if I'm going through things you've already thought of. Shouldn't the following logic cover both cases without getting stuck?

1. Do I have an IP address? 
  - Yes:
    - Continue on
  - No:
    - Have 45 seconds elapsed?
      - Yes:
        - Exit with Error: failed to boot to network!
      - No:
         - Wait 1 second
         - Loop back to "Do I have an IP address?"
2. Can I ssh in?
  - Yes:
    - Continue on
  - No:
    - Have 300 seconds elapsed?
      - Yes:
        - Exit with Error: got network, but no ssh!
      - No:
         - Wait 1 second
         - Loop back to "Can I ssh in?"

This has a minimal delay in the optimal case, fails after 45 seconds in the "totally failed" case (that probably could be brought down to 30, if it's useful — we should get at least that far very quickly, right?), and gives up to five minutes for ssh to initialize.

**Comment by kushaldas** [comment]
Will do this.
